### PR TITLE
Use many-to-one edges in pagination planning

### PR DESCRIPTION
**Pull request by bojanserafimov**
Make sure there is no unique filter on query nodes reachable from the pagination node by many-to-one edges.

### CODE REVIEW

File: graphql_compiler/cost_estimation/analysis.py [hunk 1]
```diff
@@ -125,10 +126,28 @@ def get_distinct_result_set_estimates(
...
 
-    # TODO(bojanserafimov): If there's a many-to-one edge from A to B in the query, the
-    #                       distinct result set estimate for B cannot be greater than
-    #                       the one for A. Taking this into account would make the results
-    #                       more accurate.
+    single_destination_traversals = set()
+    for location, _ in query_metadata.registered_locations:
+        if len(location.query_path) > 1:
+            from_path = location.query_path[:-1]
+            to_path = location.query_path
```
Comment (obi1kenobi) [inline graphql_compiler/cost_estimation/analysis.py]: These are not the same type -- one is the entire query path, and one is just the last element in it. That doesn't seem intended..?

File: graphql_compiler/tests/snapshot_tests/test_query_pagination.py [hunk 2]
```diff
@@ -108,6 +108,53 @@ def test_pagination_planning_unique_filter(self) -> None:
...
+        number_of_pages = 10
+        analysis = analyze_query_string(schema_info, query)
+        pagination_plan, warnings = get_pagination_plan(analysis, number_of_pages)
+
+        # This is a white box test. We check that we don't paginate on the root when it has a
+        # unique filter on its many-to-one neighbor. A better plan is to paginate on a
+        # different vertex, but that is not implemented.
+        expected_plan = PaginationPlan(tuple())
+        expected_warnings: Tuple[PaginationAdvisory, ...] = tuple()
+        self.assertEqual([w.message for w in expected_warnings], [w.message for w in warnings])
```
Comment (obi1kenobi) [inline graphql_compiler/tests/snapshot_tests/test_query_pagination.py]: Warnings are something else in Python, let's stick with the "advisory" terminology: https://docs.python.org/3/library/warnings.html

File: graphql_compiler/tests/snapshot_tests/test_query_pagination.py [hunk 3]
```diff
@@ -108,6 +108,53 @@ def test_pagination_planning_unique_filter(self) -> None:
...
+            type_equivalence_hints=type_equivalence_hints,
+            schema_graph=schema_graph,
+            statistics=statistics,
+            pagination_keys=pagination_keys,
+            uuid4_fields=uuid4_fields,
+            edge_constraints=edge_constraints,
+        )
+
+        query = QueryStringWithParameters(
+            """{
```
Comment (obi1kenobi) [inline graphql_compiler/tests/snapshot_tests/test_query_pagination.py]: The formatting here is a bit weird.

File: graphql_compiler/cost_estimation/analysis.py (hunk 1 above)
Comment (bojanserafimov) [reply graphql_compiler/cost_estimation/analysis.py]: It's all but the last element

File: graphql_compiler/cost_estimation/analysis.py [hunk 4]
```diff
@@ -130,10 +130,32 @@ def get_distinct_result_set_estimates(
...
+            # TODO(bojanserafimov): We currently ignore FoldScopeLocations. However, a unique
+            #                       filter on a FoldScopeLocation also uniquely filters the
+            #                       enclosing scope.
+            continue
+
+        if len(location.query_path) > 1:
+            from_path = location.query_path[:-1]
+            to_path = location.query_path
+            edge_direction, edge_name = get_edge_direction_and_name(location.query_path[-1])
+            edge_constraints = schema_info.edge_constraints.get(edge_name, EdgeConstraint(0))
```
Comment (pmantica1) [inline graphql_compiler/cost_estimation/analysis.py]: I am not sure what this syntax `EdgeConstraint(0)` does, but this seems sketchy. Do you mind explaining or fixing?

File: graphql_compiler/cost_estimation/analysis.py [hunk 5]
```diff
@@ -130,10 +130,32 @@ def get_distinct_result_set_estimates(
...
+            if edge_direction == "in":
+                from_path, to_path = to_path, from_path
+
+            if EdgeConstraint.AtMostOneDestination in edge_constraints:
+                single_destination_traversals.add((from_path, to_path))
+            if EdgeConstraint.AtMostOneSource in edge_constraints:
+                single_destination_traversals.add((to_path, from_path))
+
+    for _ in single_destination_traversals:  # This is on purpose (TODO explain why)
```
Comment (pmantica1) [inline graphql_compiler/cost_estimation/analysis.py]: Can you address this TODO? I understand what you are doing here so don't feel the need to go into a lot of detail. Maybe a link to a similar algorithm would be sufficient.

File: graphql_compiler/cost_estimation/analysis.py (hunk 4 above)
Comment (bojanserafimov) [reply graphql_compiler/cost_estimation/analysis.py]: Flags are represented with bits. 0 means all bits unset. Can't find another way to get this value.

**Pull request merged**